File: app/src/main/java/gm/tieba/tabswitch/hooker/eliminate/FrsPageFilter.java
package gm.tieba.tabswitch.hooker.eliminate;

import androidx.annotation.NonNull;

import java.util.List;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedHelpers;
import gm.tieba.tabswitch.XposedContext;
import gm.tieba.tabswitch.hooker.IHooker;

public class FrsPageFilter extends XposedContext implements IHooker, RegexFilter {

    @NonNull
    @Override
    public String key() {
        return "frs_page_filter";
    }

    @Override
    public void hook() throws Throwable {
        XposedHelpers.findAndHookMethod("tbclient.FrsPage.PageData$Builder", sClassLoader, "build", boolean.class, new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(final MethodHookParam param) throws Throwable {
                filterPageData(param.thisObject);
            }
        });

        XposedHelpers.findAndHookMethod("tbclient.ThreadList.PageData$Builder", sClassLoader, "build", boolean.class, new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(final MethodHookParam param) throws Throwable {
                filterPageData(param.thisObject);
            }
        });
    }

    private void filterPageData(Object pageData) {
        List<?> feedList = (List<?>) XposedHelpers.getObjectField(pageData, "feed_list");
        if (feedList == null) return;
        final var pattern = getPattern();
        feedList.removeIf(
                o -> {
                    Object currFeed = XposedHelpers.getObjectField(o, "feed");
                    if (currFeed != null) {
                        List<?> businessInfo = (List<?>) XposedHelpers.getObjectField(currFeed, "business_info");
                        for (var feedKV : businessInfo) {
                            String currKey = XposedHelpers.getObjectField(feedKV, "key").toString();
                            if (currKey.equals("title") || currKey.equals("abstract")) {
                                String str = XposedHelpers.getObjectField(feedKV, "value").toString();
                                if (pattern.matcher(str).find()) {
                                    return true;
                                }
                            }
                        }
                    }
                    return false;
                }
        );
    }
}

File: app/src/main/java/gm/tieba/tabswitch/hooker/eliminate/FrsPageFilter.kt
package gm.tieba.tabswitch.hooker.eliminate

import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XposedHelpers
import gm.tieba.tabswitch.XposedContext
import gm.tieba.tabswitch.hooker.IHooker

class FrsPageFilter : XposedContext(), IHooker, RegexFilter {
    override fun key(): String {
        return "frs_page_filter"
    }

    @Throws(Throwable::class)
    override fun hook() {
        hookBeforeMethod(
            "tbclient.FrsPage.PageData\$Builder",
            "build", Boolean::class.javaPrimitiveType
        ) { param-> filterPageData(param.thisObject) }
        hookBeforeMethod(
            "tbclient.ThreadList.PageData\$Builder",
            "build", Boolean::class.javaPrimitiveType
        ) { param-> filterPageData(param.thisObject) }
    }

    private fun filterPageData(pageData: Any) {
        val feedList = XposedHelpers.getObjectField(pageData, "feed_list") as? MutableList<*>
        val pattern = getPattern() ?: return

        feedList?.removeIf { feedItem ->
            val currFeed = XposedHelpers.getObjectField(feedItem, "feed")

            currFeed?.let { feed ->
                val businessInfo = XposedHelpers.getObjectField(feed, "business_info") as? List<*>

                businessInfo?.any { feedKV ->
                    val currKey = XposedHelpers.getObjectField(feedKV, "key").toString()
                    if (currKey == "title" || currKey == "abstract") {
                        val str = XposedHelpers.getObjectField(feedKV, "value").toString()
                        pattern.matcher(str).find()
                    } else {
                        false
                    }
                } ?: false
            } ?: false
        }
    }
}
